File: edu/valid_palindrome.cpp
#include <string>
#include <iostream>
#include <cctype>

using namespace std;

class Solution
{
public:
    bool isPalindrome(string s)
    {
        for (int i = 0, j = s.size() - 1; i < j;)
        {

            if (!isalnum(s[i]))
            {
                ++i;
                continue;
            }
            if (!isalnum(s[j]))
            {
                --j;
                continue;
            }

            if (tolower(s[i]) != tolower(s[j]))
                return false;

            ++i;
            --j;
        }
        return true;
    }
};

int main()
{
    string str = "Was it a car or a cat I saw?";
    Solution s;
    cout << s.isPalindrome(str);
    return 0;
}

File: java/src/com/leetcode/refresh/ValidPalindrome.java
package com.leetcode.refresh;

public class ValidPalindrome {

    public static void main(String[] args) {
        System.out.println(
                new ValidPalindrome().isPalindrome(
                        "Was it a car or a cat I saw?"
                )
        );
    }

    public boolean isPalindrome(String s) {
        int left = 0;
        int right = s.length() - 1;

        while (left < right) {
            if (!Character.isLetterOrDigit(s.charAt(left))) {
                left++;
                continue;
            }

            if (!Character.isLetterOrDigit(s.charAt(right))) {
                right--;
                continue;
            }

            if(Character.toLowerCase(s.charAt(left)) != Character.toLowerCase(s.charAt(right))) {
                return false;
            }
            left++;
            right--;
        }

        return true;

    }
}
